### PR TITLE
Fix ARMViolations for Microsoft.Resources

### PR DESCRIPTION
**Pull request by zjpjack**
Fix ARMViolations for Microsoft.Resources

### CODE REVIEW

File: specification/resources/resource-manager/Microsoft.Resources/stable/2018-05-01/resources.json [hunk 1]
```diff
@@ -34,6 +34,31 @@
     }
```
Comment (zjpjack) [inline specification/resources/resource-manager/Microsoft.Resources/stable/2018-05-01/resources.json]: In this file, when we use autorest to check ARM violations, we always have ARMViolations for "TagDetails". The error is that we must have name and type properties for it, and "TagDetails" has extra properties ['tagName', 'count', 'values']. I checked resources.json for 2018-02-01, the definition for "TagDetails" is the same, but there is no such errors in autorest. Do you know how I can address it?

**Comment by zjpjack** [comment]
@lmazuel Can you please take a look? Thanks

**Comment by lmazuel** [comment]
Hi @zjpjack somehow I missed your PR in my notification, which is unusual :(. Taking a look right now.

File: specification/resources/resource-manager/Microsoft.Resources/stable/2018-05-01/resources.json [hunk 2]
```diff
@@ -3000,9 +3028,15 @@
           "description": "The ID of the resource group."
         },
         "name": {
+          "readOnly": true,
           "type": "string",
           "description": "The name of the resource group."
         },
+        "type": {
```
Comment (lmazuel) [inline specification/resources/resource-manager/Microsoft.Resources/stable/2018-05-01/resources.json]: Just check some SDK recording, I don't see `type`

```json
{
	"id": "/subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/test_mgmt_resource_test_resource_groups457f1050",
	"name": "test_mgmt_resource_test_resource_groups457f1050",
	"location": "westus",
	"tags": {
		"tag1": "value1"
	},
	"properties": {
		"provisioningState": "Succeeded"
	}
}
```

File: specification/resources/resource-manager/Microsoft.Resources/stable/2018-05-01/resources.json [hunk 3]
```diff
@@ -3167,6 +3201,7 @@
           "description": "The list of tag values."
         }
       },
+      "x-ms-azure-resource": true,
```
Comment (lmazuel) [inline specification/resources/resource-manager/Microsoft.Resources/stable/2018-05-01/resources.json]: TagDetails is not a resource

**Review by lmazuel — CHANGES_REQUESTED** [review]

File: specification/resources/resource-manager/Microsoft.Resources/stable/2016-06-01/subscriptions.json [hunk 4]
```diff
@@ -312,6 +337,49 @@
...
+      "description": "Microsoft.Resources operation",
+      "type": "object",
+      "properties": {
+        "name": {
+          "description": "Operation name: {provider}/{resource}/{operation}",
+          "type": "string"
+        },
+        "display": {
+          "description": "The object that represents the operation.",
+          "properties": {
```
Comment (lmazuel) [inline specification/resources/resource-manager/Microsoft.Resources/stable/2016-06-01/subscriptions.json]: Missing required "description":
https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/proxy-api-reference.md#exposing-available-operations

File: specification/resources/resource-manager/Microsoft.Resources/stable/2016-09-01/links.json [hunk 5]
```diff
@@ -291,6 +321,49 @@
...
+      "description": "Microsoft.Resources operation",
+      "type": "object",
+      "properties": {
+        "name": {
+          "description": "Operation name: {provider}/{resource}/{operation}",
+          "type": "string"
+        },
+        "display": {
+          "description": "The object that represents the operation.",
+          "properties": {
```
Comment (lmazuel) [inline specification/resources/resource-manager/Microsoft.Resources/stable/2016-09-01/links.json]: Missing required "description":
https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/proxy-api-reference.md#exposing-available-operations

File: specification/resources/resource-manager/Microsoft.Resources/stable/2018-05-01/resources.json [hunk 6]
```diff
@@ -3373,6 +3408,49 @@
...
+      "description": "Microsoft.Resources operation",
+      "type": "object",
+      "properties": {
+        "name": {
+          "description": "Operation name: {provider}/{resource}/{operation}",
+          "type": "string"
+        },
+        "display": {
+          "description": "The object that represents the operation.",
+          "properties": {
```
Comment (lmazuel) [inline specification/resources/resource-manager/Microsoft.Resources/stable/2018-05-01/resources.json]: Missing required "description":
https://github.com/Azure/azure-resource-manager-rpc/blob/master/v1.0/proxy-api-reference.md#exposing-available-operations

File: specification/resources/resource-manager/Microsoft.Resources/stable/2016-06-01/subscriptions.json (hunk 4 above)
Comment (zjpjack) [reply specification/resources/resource-manager/Microsoft.Resources/stable/2016-06-01/subscriptions.json]: You mean the display.description?

File: specification/resources/resource-manager/Microsoft.Resources/stable/2018-05-01/resources.json [hunk 7]
```diff
@@ -2785,9 +2810,15 @@
           "description": "The ID of the deployment."
         },
         "name": {
+          "readOnly": true,
           "type": "string",
           "description": "The name of the deployment."
         },
+        "type": {
```
[the remark this answers is not in the document]
Comment (zjpjack) [reply specification/resources/resource-manager/Microsoft.Resources/stable/2018-05-01/resources.json]: According to Autorest, if there is no "type", then we get the following error 
ERROR (RequiredPropertiesMissingInResourceModel/R2020/ARMViolation): Model definition 'DeploymentExtended' must have the properties 'name', 'id' and 'type' in its hierarchy and these properties must be marked as readonly.
Any suggestion for this?

File: specification/resources/resource-manager/Microsoft.Resources/stable/2018-05-01/resources.json (hunk 3 above)
Comment (zjpjack) [reply specification/resources/resource-manager/Microsoft.Resources/stable/2018-05-01/resources.json]: You are right, removed this.

**Comment by zjpjack** [comment]
@lmazuel Replied your comments. Can you please take a look?

File: specification/resources/resource-manager/Microsoft.Resources/stable/2016-06-01/subscriptions.json (hunk 4 above)
Comment (lmazuel) [reply specification/resources/resource-manager/Microsoft.Resources/stable/2016-06-01/subscriptions.json]: Yes

Comment (zjpjack) [reply specification/resources/resource-manager/Microsoft.Resources/stable/2016-06-01/subscriptions.json]: For display.description, I already had it. It is above display.properties

File: specification/resources/resource-manager/Microsoft.Resources/stable/2018-05-01/resources.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (lmazuel) [reply specification/resources/resource-manager/Microsoft.Resources/stable/2018-05-01/resources.json]: The most important is that the Swagger represent the absolute truth about the server. This is indeed an ARM violation, the linter is right that this call does not respect ARM.
Correct to handle this is to send an email to @ravbhatnagar to describe the issue and ask for a ARM suppression rule. So at least the non-conformance is known and documented. Adding type here just shut off the alarm and doesn't solve anything.

**Comment by lmazuel** [comment]
@zjpjack Added comments. Thanks!

**Comment by zjpjack** [comment]
@lmazuel For those comments about missing discription for display, I already had them in this PR, it is just below the display

File: specification/resources/resource-manager/Microsoft.Resources/stable/2016-06-01/subscriptions.json [hunk 8]
```diff
@@ -312,6 +337,49 @@
...
+    "Operation": {
+      "description": "Microsoft.Resources operation",
+      "type": "object",
+      "properties": {
+        "name": {
+          "description": "Operation name: {provider}/{resource}/{operation}",
+          "type": "string"
+        },
+        "display": {
+          "description": "The object that represents the operation.",
```
Comment (zjpjack) [inline specification/resources/resource-manager/Microsoft.Resources/stable/2016-06-01/subscriptions.json]: Do you mean the description here? I already had it for display.

Comment (lmazuel) [reply specification/resources/resource-manager/Microsoft.Resources/stable/2016-06-01/subscriptions.json]: "description" the Swagger keyword, and "description" the JSON key in ARM answer are too different things. You have a Swagger "description" of "display", but "display" can contain a property called "description" which has to be defined:
```json
  "display": {
    "provider": "{Name of the provider for display purposes}",
    "resource": "{Name of the resource type for display purposes}",
    "operation": "{Name of the operation for display purposes}",
    "description": "{Description of the operation for display purposes}"
  },
```

Example:
https://github.com/Azure/azure-rest-api-specs/blob/57d827ec31e84a77fa35dfe6bdd048eb84916918/specification/storage/resource-manager/Microsoft.Storage/stable/2018-07-01/storage.json#L581-L601

Comment (zjpjack) [reply specification/resources/resource-manager/Microsoft.Resources/stable/2016-06-01/subscriptions.json]: Thanks, it makes sense to me :)

File: specification/resources/resource-manager/Microsoft.Resources/stable/2018-05-01/resources.json (hunk 1 above)
Comment (zjpjack) [reply specification/resources/resource-manager/Microsoft.Resources/stable/2018-05-01/resources.json]: @lmazuel I will leave all these suppression to @ifeoluwaokunoren. In this PR, I will fix others, he will send a new PR for suppression.

**Review by lmazuel — APPROVED** [review]

File: specification/resources/resource-manager/Microsoft.Resources/stable/2018-05-01/resources.json (hunk 7 above)
[the remark this answers is not in the document]
Comment (zjpjack) [reply specification/resources/resource-manager/Microsoft.Resources/stable/2018-05-01/resources.json]: I will let @ravbhatnagar knows this suppression, he or I will send another PR for these suppression rules.

File: specification/resources/resource-manager/Microsoft.Resources/stable/2018-05-01/resources.json (hunk 1 above)
Comment (zjpjack) [reply specification/resources/resource-manager/Microsoft.Resources/stable/2018-05-01/resources.json]: I will let @ravbhatnagar knows this suppression, he or I will send another PR for these suppression rules.

**Pull request merged**